### PR TITLE
Implement new UI design, add story and unit tests

### PR DESCRIPTION
**Pull request by deniskaber**
Part of https://github.com/metabase/metabase/issues/28687

### Description

Implement new UI design, add story and unit tests
Design - https://www.figma.com/file/68g8S9L4cMesbctGfCySl7/De-crazy-the-expression-editor-function-helper-popover?node-id=9%3A153&t=hFinhfY6ezqtXbBz-4

### How to verify

- yarn storybook
- Open `Query Builder/ExpressionWidget` story

### Demo

https://user-images.githubusercontent.com/7983744/222142134-33f19527-1a18-46a3-b34d-c8b5b7b1d3b4.mov

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/28776)

### CODE REVIEW

File: frontend/src/metabase-lib/metadata/Database.ts [hunk 1]
```diff
@@ -40,6 +40,16 @@ class DatabaseInner extends Base {
...
+  constructor(object) {
+    super(object);
+
+    if (window.IS_STORYBOOK) {
+      for (const property in object) {
+        this[property] = object[property];
+      }
+    }
+  }
+
```
Comment (kulyk) [inline frontend/src/metabase-lib/metadata/Database.ts:52]: Not sure we should make `metabase-lib` dependant on Storybook or `window`
What are you trying to achieve here?

File: frontend/src/metabase-lib/metadata/Table.ts [hunk 2]
```diff
@@ -35,6 +35,16 @@ class TableInner extends Base {
...
+  constructor(object) {
+    super(object);
+
+    if (window.IS_STORYBOOK) {
+      for (const property in object) {
+        this[property] = object[property];
+      }
+    }
+  }
+
```
Comment (kulyk) [inline frontend/src/metabase-lib/metadata/Table.ts:47]: Same feedback as for the similar change in `Database`

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import type { ComponentStory } from "@storybook/react";
+import { useArgs } from "@storybook/client-api";
+
+import styled from "@emotion/styled";
+import stateFixture from "__support__/sample_database_fixture.json";
+import { getMetadata } from "metabase/selectors/metadata";
+import { State } from "metabase-types/store";
+import StructuredQuery from "metabase-lib/queries/StructuredQuery";
+import ExpressionWidget from "./ExpressionWidget";
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx:11]: Please make it follow our [import ordering guidelines](https://www.metabase.com/docs/latest/developers-guide/frontend.html#javascript-conventions)

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+const state = stateFixture as unknown as State;
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx:13]: We're planning to eventually get rid of `sample_database_fixture`. Can we compose a meaningful dataset using helpers from `metabase-types/api/mock`? 

<details>
<summary>Example usage</summary>

```js
import {
  createMockDatabase,
  createMockField,
  createMockTable,
} from "metabase-types/api/mocks";
import { TYPE } from "metabase-lib/types/constants";

const field1 = createMockField({ id: 1, name: "ID", semantic_type: TYPE.PK });
const field2 = createMockField({ id: 2, name: "Name" }); 

const table1 = createMockTable({ id: 1, name: "Table", fields: [ field1, field2 ] });

const database = createMockDatabase({ id: 1, tables: [ table1 ] });
```

</details>

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+const metadata = getMetadata(state);
+const table = metadata.table(1);
+const query = new StructuredQuery(table?.question(), {
+  type: "query",
+  database: table?.db_id,
+  query: {
+    "source-table": table?.id,
+  },
+});
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.stories.tsx:22]: Think you can do `table?.question().query()` here instead?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+function setup(additionalProps?: Partial<ExpressionWidgetProps>) {
+  const TEST_QUERY = {
+    type: "query",
+    database: SAMPLE_DATABASE.id,
+    query: {
+      "source-table": ORDERS.id,
+    },
+  };
+  const query = new StructuredQuery(ORDERS.question(), TEST_QUERY);
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:44]: Can you do `ORDERS.question().query()` here instead?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    expect(
+      screen.getByText(
+        "You can reference columns here in functions or equations, like: floor([Price] - [Discount]).",
+      ),
+    ).toBeInTheDocument();
+  });
+});
+
+function setup(additionalProps?: Partial<ExpressionWidgetProps>) {
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:36]: nit: I think a more common pattern across Metabase is to define default props when restructuring an argument:

```tsx
function setup({
  reportTimezone = "UTC",
  onChangeExpression = jest.fn(),
  ...props
}: Partial<ExpressionWidgetProps>) {
  ...
}
```

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    expect(screen.getByText("Cancel")).toBeInTheDocument();
+    expect(screen.getByText("Done")).toBeInTheDocument();
+  });
+
+  it("should render help icon with tooltip", () => {
+    setup();
+
+    const iconEl = screen.getByRole("img");
+    expect(iconEl).toBeInTheDocument();
+    expect(iconEl).toHaveClass("Icon-info");
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:24]: We can use a `getIcon` helper from `__support__/ui` to avoid the CSS class assertion:

```jsx
import { getIcon } from "__support__/ui"

const icon = getIcon("info")
expect(icon).toBeInTheDocument()
userEvent.hover(icon)
```

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import React from "react";
+import { render, screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import { ORDERS, SAMPLE_DATABASE } from "__support__/sample_database_fixture";
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:4]: As in the [Storybook file comment](https://github.com/metabase/metabase/pull/28776#discussion_r1121740394), it'd be awesome if we can avoid using the sample database fixture

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 10]
```diff
@@ -52,8 +55,19 @@ const ExpressionWidget = (props: ExpressionWidgetProps): JSX.Element => {
 
   return (
     <ExpressionWidgetWrapper>
-      <div className="p2">
-        <div className="h5 text-uppercase text-light text-bold">{t`Expression`}</div>
+      <div className="p3 pb4">
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:58]: For new components, we try to remove the usage of class utilities and favor using `emotion` instead. Could you extract any components that have the className changes to `ExpressionWidget.style.ts` instead?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import StructuredQuery from "metabase-lib/queries/StructuredQuery";
+import ExpressionWidget, { ExpressionWidgetProps } from "./ExpressionWidget";
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:6]: I think lately we've been adopting `import type` for types. I think you should extract the type import to another line what do you think @kulyk I think you also use `import type` a lot too.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx (hunk 8 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:24]: 👆 note that the path has changed since - https://github.com/metabase/metabase/pull/28749

File: frontend/test/metabase/scenarios/custom-column/cc-help-text.cy.spec.js [hunk 12]
```diff
@@ -34,7 +34,7 @@ describe("scenarios > question > custom column > help text", () => {
     cy.findByText("round([Temperature])");
 
     // Click outside of formula field instead of blur
-    cy.findByText(/Field formula/i).click();
+    cy.findByText(/Expression/i).click();
```
Comment (WiNloSt) [inline frontend/test/metabase/scenarios/custom-column/cc-help-text.cy.spec.js:37]: would it be possible to find this element using a simple string? [Ref](https://www.notion.so/metabase/43-Update-Style-Guide-2bcbc77ff4f34c9bb74a6623db22c9d6?pvs=4#9d35175600c549cc91f8bfcce5bdafd3)

File: frontend/src/metabase-lib/metadata/Database.ts [hunk 13]
```diff
@@ -181,36 +191,6 @@ class DatabaseInner extends Base {
...
-   * @param {string} name
-   * @param {?string} description
-   * @param {Table[]} tables
-   * @param {Schema[]} schemas
-   * @param {Metadata} metadata
-   * @param {boolean} auto_run_queries
-   */
-
-  /* istanbul ignore next */
-  _constructor(
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase-lib/metadata/Database.ts:197]: Not part of JavaScript. Can't find any traces of this too, I think it can be removed

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 14]
```diff
@@ -52,79 +63,70 @@ const ExpressionWidget = (props: ExpressionWidgetProps): JSX.Element => {
...
-            <ExternalLink
-              className="link"
-              target="_blank"
-              href={MetabaseSettings.docsUrl(
-                "questions/query-builder/expressions",
-              )}
-            >{t`Learn more`}</ExternalLink>
-          </p>
-        </div>
+      <ExpressionWidgetExpressionFieldWrapper ref={helpTextTargetRef}>
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:66]: nit: Want to hear @mazameli 's opinion about this. The way the new popover is opening is to target around this element (see the highlighted green padding)

Screenshot 2023-03-02 at 1 00 25 PM

This makes the popover open a bit further from the custom editor input box and make the other text appear above the popover which is a little unpleasant to me
Screenshot 2023-03-02 at 1 00 34 PM


Compare to the design which has the popover open closer to the custom editor input box.
image

In my opinion, the popover should open closer to the input box similar to the design.

What do you guys think?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.styled.tsx [hunk 15]
```diff
@@ -1,5 +1,80 @@
...
+`;
+
+export const ExpressionWidgetStyledIcon = styled(Icon)`
+  width: 10px;
+  height: 10px;
+`;
+
+export const ExpressionWidgetInput = styled(Input)`
+  input {
+    border-radius: 4px;
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.styled.tsx:46]: @mazameli Are we moving towards using `border-radius: 4px` in input boxes? This matches the spec in the [design here](https://www.figma.com/file/68g8S9L4cMesbctGfCySl7/De-crazy-the-expression-editor-function-helper-popover?node-id=0%3A1&t=7VTlgGbjJm0Za6is-0).

But I feel like we still use `8px` in a lot of other places. e.g. dashboard parameter
image


Found this PR https://github.com/metabase/metabase/pull/23840 which standardizes input box radius from 4px to 8px too.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 16]
```diff
@@ -1,16 +1,27 @@
...
+  ExpressionWidgetExpressionFieldWrapper,
+  ExpressionWidgetFieldTitle,
+  ExpressionWidgetFieldWrapper,
+  ExpressionWidgetFooter,
+  ExpressionWidgetIconWrapper,
+  ExpressionWidgetInput,
+  ExpressionWidgetRemoveLink,
+  ExpressionWidgetStyledIcon,
+  ExpressionWidgetWrapper,
+} from "metabase/query_builder/components/expressions/ExpressionWidget.styled";
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:16]: Should this be a local import?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+  });
+
+  it("should render help icon with tooltip", () => {
+    setup({});
+
+    const icon = getIcon("info");
+    expect(icon).toBeInTheDocument();
+    userEvent.hover(icon);
+
+    userEvent.hover(icon);
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:27]: Do we need to hover twice?

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+const onChangeExpressionMock = jest.fn();
+function setup({
+  name,
+  expression,
+  query = createMockQueryFoxExpressions() as StructuredQuery,
+  reportTimezone = "UTC",
+  onChangeExpression = onChangeExpressionMock,
+  ...restProps
+}: Partial<ExpressionWidgetProps>) {
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:45]: Should we consider making a new `onChangeExpressionMock` per each `setup` call? Otherwise, we'd need to reset the mock between tests if we want to assert on its calls; it could be easy to forget

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+      ),
+    ).toBeInTheDocument();
+  });
+});
+
+const onChangeExpressionMock = jest.fn();
+function setup({
+  name,
+  expression,
+  query = createMockQueryFoxExpressions() as StructuredQuery,
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:41]: Seems like we're calling `metabase-lib/metadata/Table's` `query` method in `createMockQueryFoxExpressions`
AFAIR, it always returns a structured query, can we avoid type-casting?

File: frontend/src/metabase/query_builder/tests/common.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+  object: Record<string, unknown>,
+): T => {
+  const instance = new Class(object);
+
+  for (const property in object) {
+    (instance as any)[property] = object[property];
+  }
+
+  return instance;
+};
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/tests/common.ts:28]: I believe we need to figure out a way to avoid this, I'll try a few things locally

**Review by WiNloSt — COMMENTED** [review]
Have some design questions. Otherwise, this looks fine.

File: frontend/src/metabase/query_builder/tests/common.ts (hunk 20 above)
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: let me know if you came up with anything. I've tried, and it seems like the constructor here is doing its work properly: https://github.com/metabase/metabase/blob/65ab08c4266a3a359dec776b0bfc8669bb9b8457/frontend/src/metabase-lib/metadata/Base.ts#L9-L11

I logged that `this[property]`, and they seem to have values, but if I log `this` and observe its value at runtime, some of the values will be gone. It seems as if they're overridden somehow 🤔

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx (hunk 16 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:16]: @deniskaber why do you use names like this? We usually do not make styled components name s so unique. E.g. https://github.com/metabase/metabase/blob/c9fbddc2e8c9cf86d8c100c1237bd12deb9a72f5/frontend/src/metabase/models/components/ModelDetailPage/ModelActionDetails/ModelActionListItem.styled.tsx#L79

File: frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield/suggest.js [hunk 21]
```diff
@@ -132,16 +132,20 @@ export function suggest({
...
+    const segmentsSuggestions = query.table().segments?.map(segment => ({
+      type: "segments",
+      name: segment.name,
+      text: formatSegmentName(segment),
+      index: targetOffset,
+      icon: "segment",
+      order: 3,
+    }));
+
+    if (Array.isArray(segmentsSuggestions)) {
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield/suggest.js:145]: It's also not clear to me, spreading `undefined` should work without issues

File: frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield/suggest.js [hunk 22]
```diff
@@ -132,16 +132,20 @@ export function suggest({
...
-        type: "segments",
-        name: segment.name,
-        text: formatSegmentName(segment),
-        index: targetOffset,
-        icon: "segment",
-        order: 3,
-      })),
-    );
+
+    const segmentsSuggestions = query.table().segments?.map(segment => ({
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield/suggest.js:136]: Also not clear

File: frontend/src/metabase/query_builder/tests/common.ts (hunk 20 above)
Comment (kulyk) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: Yeah, something weird is going on. IMO if we can't find a cleaner way to fix this, I'd consider removing `ExpressionWidget` storybook file :( We have plenty of magic in the codebase already, and I'm worried about adding more. @WiNloSt @ranquild WDYT?

Comment (ranquild) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: I think it should be possible to use the helper from https://github.com/metabase/metabase/pull/28836 in storybook. Hopefully no metabase-lib changes would be required. @deniskaber could you take a look?

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: I think the benefit of having a storybook to play with, especially when modifying the component, outweighs the quirkiness we have here. I'm inclined to leave the hack here.

Comment (kulyk) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: Rendering components dealing with `metabase-lib` stuff can be complicated. Maybe we could eventually refactor `ExpressionWidget` to be independent from `metabase-lib`; adding it to Storybook would be straightforward in this case

Comment (ranquild) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: @kulyk btw WDYT about that?

Comment (kulyk) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: @ranquild sorry, we pulled a few different threads here, so I'm not sure I understand what you are referring to

> I think it should be possible to use the helper from #28836 in storybook. Hopefully, no metabase-lib changes would be required.

That might work; it'd be great if we could give it a shot

> I think the benefit of having a storybook to play with, especially when modifying the component, outweighs the quirkiness we have here. I'm inclined to leave the hack here.

If we keep these hacks/workarounds, can we be sure that components in Storybook would behave the same way as in the core app? I personally find a solid RTL test suite much more helpful when modifying a component; the tests would immediately highlight changes in behavior.

I'm probably biased because I don't use Storybook often enough here. So if the rest of the team finds it useful — let's keep it. I'd only reconsider where that code should live (probably something more top-level than the QB directory)

Comment (ranquild) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: I hope that no hacks would be needed if the helper is used. If hacks are still required, let's discuss this again. @kulyk @deniskaber @WiNloSt

Comment (WiNloSt) [reply frontend/src/metabase/query_builder/tests/common.ts:28]: > I'm probably biased because I don't use Storybook often enough here. So if the rest of the team finds it useful — let's keep it. I'd only reconsider where that code should live (probably something more top-level than the QB directory)

I used to use Storybook often, but the way some components in metabase are structured makes it hard to render in storybook or even impossible for some components. Ideally, we should be able to render any pure UI in storybook, but we seem to use our entity loaders and redux here and there.

Usually, I would run Storybook to develop any pure UI or styling change, but I have never been able to do that here 😓 

re:
> I hope that no hacks would be needed if the helper is used. If hacks are still required, let's discuss this again.

And 

> If we keep these hacks/workarounds, can we be sure that components in Storybook would behave the same way as in the core app?

I also agree with you guys, so let's see if we could get rid of this workaround.

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx (hunk 19 above)
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:41]: Yeah, `Table.prototype.query` should always return a `StructuredQuery`. Maybe we should explicitly specify the return types in `Table's` method instead, so we don't need to cast types elsewhere

**Review by ranquild — COMMENTED** [review]
@deniskaber I don't think this PR gets merged with the way the storybook is implemented. Could I ask you to move the storybook to a separate PR so we can discuss it there? I think it's wrong that we've been discussing it for so much time and not the feature instead. cc @kulyk @WiNloSt

**Comment by kulyk** [comment]
> I don't think this PR gets merged with the way the storybook is implemented. Could I ask you to move the storybook to a separate PR so we can discuss it there? I think it's wrong that we've been discussing it for so much time and not the feature instead

@ranquild good points, I'm on board with the plan 👍

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    }),
+  });
+
+  const metadata = getMetadata(state);
+  const query = metadata.table(ORDERS_ID)?.query();
+
+  return query;
+};
+
+const onChangeExpressionMock = jest.fn();
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:54]: I think we usually return a mock function from `setup`, like:
```
const setup() {
  const onChangeExpression = jest.fn();

  renderWithProviders(...);

  return { onChangeExpression }; 
});
```

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.styled.tsx [hunk 24]
```diff
@@ -1,5 +1,73 @@
...
+
+export const StyledFieldTitleIcon = styled(Icon)`
+  width: 10px;
+  height: 10px;
+`;
+
+export const Divider = styled.div`
+  width: 100%;
+  height: 1px;
+  background-color: #eeecec;
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.styled.tsx:46]: We usually don't use custom colors because they cannot be changed in the admin UI. I think it should be something like `bg-light` or other `bg-` variants.

**Review by ranquild — COMMENTED** [review]
@deniskaber the PR looks good, I have some small comments, e.g. about the use of background color, and it should be good to go

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx (hunk 23 above)
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.unit.spec.tsx:54]: I'm also fine if it was provided from the place where `setup` is called, but a global mock looks like something we don't usually do

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 25]
```diff
@@ -1,16 +1,27 @@
...
 import { t } from "ttag";
 import { isNotNull } from "metabase/core/utils/types";
-import { ExpressionWidgetWrapper } from "metabase/query_builder/components/expressions/ExpressionWidget.styled";
-import ExternalLink from "metabase/core/components/ExternalLink";
-import MetabaseSettings from "metabase/lib/settings";
-import { Expression } from "metabase-types/types/Query";
+import Tooltip from "metabase/core/components/Tooltip";
+import Button from "metabase/core/components/Button";
+import type { Expression } from "metabase-types/types/Query";
+import Input from "metabase/core/components/Input/Input";
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:7]: Usually `metabase-types` are imported after the components

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 26]
```diff
@@ -1,16 +1,27 @@
...
+  ExpressionFieldWrapper,
+  FieldTitle,
+  FieldWrapper,
+  Footer,
+  IconWrapper,
+  RemoveLink,
+  StyledFieldTitleIcon,
+  Container,
+} from "./ExpressionWidget.styled";
 import ExpressionEditorTextfield from "./ExpressionEditorTextfield";
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:22]: Usually, the `.styled` import is the last import in a file

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 27]
```diff
@@ -67,65 +89,47 @@ const ExpressionWidget = (props: ExpressionWidgetProps): JSX.Element => {
...
-                handleCommit();
-              }
-            }}
-          />
-        </div>
-      </div>
+      <Divider />
+      <Footer>
+        <ActionButtonsWrapper>
+          <Button onClick={() => onClose && onClose()}>{t`Cancel`}</Button>
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:113]: Would it make sense to disable / hide the button if `onClose` is not a function?

**Review by kulyk — APPROVED** [review]
Nice work! Thanks for addressing all the comments, left a few minor-ish items

File: frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx [hunk 28]
```diff
@@ -67,65 +89,47 @@ const ExpressionWidget = (props: ExpressionWidgetProps): JSX.Element => {
...
-        </div>
-        <div>
+          </Button>
+
           {initialName && onRemoveExpression ? (
-            <a
-              className="pr2 ml2 text-error link"
+            <RemoveLink
+              as="a"
+              href="#"
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/expressions/ExpressionWidget.tsx:122]: First of all, I don't think we've ever used `href="#"` before.
Secondly, I couldn't find a way to render this `RemoveLink`. Seems `onRemoveExpression` is only passed when we're rendering [`ExtendedOptions`](https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/frontend/src/metabase/query_builder/components/ExtendedOptions.jsx#L167).  Which only render if [`features.sort` or `features.limit` is true](https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/frontend/src/metabase/query_builder/components/ExtendedOptions.jsx#L215-L217). 

And they'll only be true when it's rendered in [`AuditQueryDetail` only](https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/enterprise/frontend/src/metabase-enterprise/audit_app/pages/AuditQueryDetail.jsx#L103) since it doesn't pass any `features` prop causing it to use the [defaultProps](https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/frontend/src/metabase/query_builder/components/GuiQueryEditor.jsx#L38-L47) instead.

The other 2 places
- https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/frontend/src/metabase/admin/datamodel/components/MetricForm/MetricForm.tsx#L65
- https://github.com/metabase/metabase/blob/2d9c6eb9e839b446ce988d4b6ded372d6dc78bbf/frontend/src/metabase/admin/datamodel/components/SegmentForm/SegmentForm.tsx#L64
don't pass `features` that has either `sort: true` or `limit: true`, so it won't render this `RemoveLink` for sure.

From that point, we can trace that we will render `AuditQueryDetail` when the path is `/admin/audit/query/:queryHash`.

But the only way to go to that page is by clicking the database in `/admin/audit/databases/all` and click on any ad-hoc query. Here, the page will crash, but we can [remove this line to circumvent this bug](https://github.com/metabase/metabase/blob/e92dc65fa94429dd5f737943d98e7a47e17ba3fa/enterprise/frontend/src/metabase-enterprise/audit_app/pages/AuditQueryDetail.jsx#L117) (I'm not sure why it crashes, but that a different story).

From there we'll finally see the `GuiQueryEditor` but it has `readOnly` props and isn't seemed to be clickable at all.
image


In conclusion, I don't think this code will ever get called. Maybe in the future, we could remove this `<RemoveLink />` from this component.

But back to the topic at hand, since I don't think this code get called, you could either leave code as is (before the change) or remove this `href="#"`.

**Review by WiNloSt — APPROVED** [review]
One comment about `href="#"` which I don't think we do that, otherwise, PR looks good. Also tested and UI looks more consistent with the rest of the app. Thank you for addressing the feedback. Great work 👍

**Review by ranquild — CHANGES_REQUESTED** [review]
Looks good but there are some items we need to polish. Left some comments, the most important one is about using a link for something that is not a link.

**Review by ranquild — APPROVED** [review]
